File: go-server/internal/handler/objectstore.go
package handler

import (
	"bytes"
	"context"
	"encoding/base64"
	"io"
	"mime"
	"net/http"
	"strings"
	"time"

	"github.com/nats-io/nats.go/jetstream"

	"nats-explorer/internal/connection"
)

type ObjectStoreHandler struct {
	Store *connection.Store
}

const objStreamPrefix = "OBJ_"

func (h *ObjectStoreHandler) store(w http.ResponseWriter, r *http.Request, timeout time.Duration) (jetstream.ObjectStore, context.Context, context.CancelFunc, bool) {
	js, err := jetStreamFor(h.Store, r)
	if err != nil {
		writeError(w, http.StatusBadRequest, err.Error())
		return nil, nil, nil, false
	}
	ctx, cancel := context.WithTimeout(r.Context(), timeout)
	os, err := js.ObjectStore(ctx, urlParam(r, "store"))
	if err != nil {
		cancel()
		writeError(w, http.StatusNotFound, err.Error())
		return nil, nil, nil, false
	}
	return os, ctx, cancel, true
}

func (h *ObjectStoreHandler) ListStores(w http.ResponseWriter, r *http.Request) {
	js, err := jetStreamFor(h.Store, r)
	if err != nil {
		writeError(w, http.StatusBadRequest, err.Error())
		return
	}

	ctx, cancel := context.WithTimeout(r.Context(), 10*time.Second)
	defer cancel()

	stores := make([]map[string]interface{}, 0)
	sl := js.ListStreams(ctx)
	for si := range sl.Info() {
		if !strings.HasPrefix(si.Config.Name, objStreamPrefix) {
			continue
		}
		stores = append(stores, map[string]interface{}{
			"bucket":            strings.TrimPrefix(si.Config.Name, objStreamPrefix),
			"description":       si.Config.Description,
			"size":              si.State.Bytes,
			"storage":           si.Config.Storage.String(),
			"sealed":            si.Config.Sealed,
			"replicas":          si.Config.Replicas,
			"chunks":            si.State.Msgs,
			"backingStreamName": si.Config.Name,
		})
	}
	if sl.Err() != nil && len(stores) == 0 {
		writeError(w, http.StatusBadGateway, sl.Err().Error())
		return
	}
	writeJSON(w, stores)
}

func (h *ObjectStoreHandler) CreateStore(w http.ResponseWriter, r *http.Request) {
	js, err := jetStreamFor(h.Store, r)
	if err != nil {
		writeError(w, http.StatusBadRequest, err.Error())
		return
	}

	var cfg struct {
		Bucket      string `json:"bucket"`
		Description string `json:"description,omitempty"`
		MaxBytes    int64  `json:"maxBytes,omitempty"`
		Storage     string `json:"storage,omitempty"`
		Replicas    int    `json:"replicas,omitempty"`
		TTL         int64  `json:"ttl,omitempty"`
	}
	if err := decodeBody(r, &cfg); err != nil {
		writeError(w, http.StatusBadRequest, err.Error())
		return
	}
	if cfg.Bucket == "" {
		writeError(w, http.StatusBadRequest, "bucket required")
		return
	}

	oscfg := jetstream.ObjectStoreConfig{Bucket: cfg.Bucket, Description: cfg.Description}
	if cfg.MaxBytes > 0 {
		oscfg.MaxBytes = cfg.MaxBytes
	}
	if cfg.Replicas > 0 {
		oscfg.Replicas = cfg.Replicas
	}
	if cfg.TTL > 0 {
		oscfg.TTL = time.Duration(cfg.TTL)
	}
	if cfg.Storage == "memory" {
		oscfg.Storage = jetstream.MemoryStorage
	}

	ctx, cancel := context.WithTimeout(r.Context(), 10*time.Second)
	defer cancel()

	if _, err := js.CreateObjectStore(ctx, oscfg); err != nil {
		writeError(w, http.StatusBadGateway, err.Error())
		return
	}
	writeJSON(w, map[string]interface{}{"success": true, "bucket": cfg.Bucket})
}

func (h *ObjectStoreHandler) DeleteStore(w http.ResponseWriter, r *http.Request) {
	js, err := jetStreamFor(h.Store, r)
	if err != nil {
		writeError(w, http.StatusBadRequest, err.Error())
		return
	}
	ctx, cancel := context.WithTimeout(r.Context(), 10*time.Second)
	defer cancel()

	if err := js.DeleteObjectStore(ctx, urlParam(r, "store")); err != nil {
		writeError(w, http.StatusBadGateway, err.Error())
		return
	}
	writeJSON(w, map[string]bool{"success": true})
}

func (h *ObjectStoreHandler) ListObjects(w http.ResponseWriter, r *http.Request) {
	os, ctx, cancel, ok := h.store(w, r, 10*time.Second)
	if !ok {
		return
	}
	defer cancel()

	objects := make([]map[string]interface{}, 0)
	list, err := os.List(ctx)
	if err != nil {
		if err == jetstream.ErrNoObjectsFound {
			writeJSON(w, objects)
			return
		}
		writeError(w, http.StatusBadGateway, err.Error())
		return
	}

	for _, info := range list {
		if info.Deleted {
			continue
		}
		objects = append(objects, map[string]interface{}{
			"name":        info.Name,
			"description": info.Description,
			"size":        info.Size,
			"chunks":      info.Chunks,
			"nuid":        info.NUID,
			"digest":      info.Digest,
			"deleted":     info.Deleted,
			"mtime":       info.ModTime.Format(time.RFC3339),
			"headers":     info.Headers,
		})
	}
	writeJSON(w, objects)
}

func (h *ObjectStoreHandler) GetObject(w http.ResponseWriter, r *http.Request) {
	os, ctx, cancel, ok := h.store(w, r, 60*time.Second)
	if !ok {
		return
	}
	defer cancel()

	objName := urlParam(r, "name")
	result, err := os.Get(ctx, objName)
	if err != nil {
		writeError(w, http.StatusNotFound, err.Error())
		return
	}
	defer result.Close()

	ctype := mime.TypeByExtension(extOf(objName))
	if ctype == "" {
		ctype = "application/octet-stream"
	}
	w.Header().Set("Content-Type", ctype)
	w.Header().Set("Content-Disposition", mime.FormatMediaType("attachment", map[string]string{"filename": objName}))
	if info, err := result.Info(); err == nil && info.Size > 0 {
		w.Header().Set("Content-Length", uintToString(info.Size))
	}
	io.Copy(w, result)
}

// PutObject accepts either a raw body (any Content-Type except JSON) or a JSON
// envelope {"data": "<base64>", "description": "..."}.
func (h *ObjectStoreHandler) PutObject(w http.ResponseWriter, r *http.Request) {
	os, ctx, cancel, ok := h.store(w, r, 5*time.Minute)
	if !ok {
		return
	}
	defer cancel()

	objName := urlParam(r, "name")
	meta := jetstream.ObjectMeta{Name: objName}

	var reader io.Reader
	if strings.HasPrefix(r.Header.Get("Content-Type"), "application/json") {
		var body struct {
			Data        string `json:"data"`
			Description string `json:"description,omitempty"`
		}
		if err := decodeBody(r, &body); err != nil {
			writeError(w, http.StatusBadRequest, err.Error())
			return
		}
		data, err := base64.StdEncoding.DecodeString(body.Data)
		if err != nil {
			writeError(w, http.StatusBadRequest, "invalid base64 data")
			return
		}
		meta.Description = body.Description
		reader = bytes.NewReader(data)
	} else {
		meta.Description = r.URL.Query().Get("description")
		reader = r.Body
	}

	info, err := os.Put(ctx, meta, reader)
	if err != nil {
		writeError(w, http.StatusBadGateway, err.Error())
		return
	}
	writeJSON(w, map[string]interface{}{
		"success": true,
		"name":    objName,
		"size":    info.Size,
		"chunks":  info.Chunks,
	})
}

func (h *ObjectStoreHandler) DeleteObject(w http.ResponseWriter, r *http.Request) {
	os, ctx, cancel, ok := h.store(w, r, 10*time.Second)
	if !ok {
		return
	}
	defer cancel()

	if err := os.Delete(ctx, urlParam(r, "name")); err != nil {
		writeError(w, http.StatusBadGateway, err.Error())
		return
	}
	writeJSON(w, map[string]bool{"success": true})
}

func extOf(name string) string {
	if i := strings.LastIndex(name, "."); i >= 0 {
		return name[i:]
	}
	return ""
}

func uintToString(n uint64) string {
	if n == 0 {
		return "0"
	}
	var buf [20]byte
	i := len(buf)
	for n > 0 {
		i--
		buf[i] = byte('0' + n%10)
		n /= 10
	}
	return string(buf[i:])
}

File: server/src/routes/objectstore.ts
import { Router } from 'express';
import { z } from 'zod';
import { connectionStore } from '../nats/connection-manager.js';

export const objectStoreRouter: Router = Router();

async function getJetStream(connId: string) {
  const nc = connectionStore.getConnection(connId);
  if (!nc) throw new Error('Not connected');
  return nc.jetstream();
}

objectStoreRouter.get('/objectstore', async (req, res) => {
  try {
    const connId = req.query.connId as string;
    if (!connId) return res.status(400).json({ error: 'connId required' });
    const nc = connectionStore.getConnection(connId);
    if (!nc) return res.status(400).json({ error: 'Not connected' });

    const jsm = await nc.jetstreamManager();
    const stores: any[] = [];
    const streams = jsm.streams.list();
    for await (const si of streams) {
      if (si.config.name.startsWith('OBJ_')) {
        const storeName = si.config.name.slice(4);
        stores.push({
          bucket: storeName,
          description: si.config.description,
          size: si.state.bytes,
          storage: si.config.storage,
          sealed: false,
          replicas: si.config.num_replicas,
          chunks: si.state.messages,
          backingStreamName: si.config.name,
        });
      }
    }
    res.json(stores);
  } catch (err: any) {
    res.status(500).json({ error: err.message });
  }
});

const objStoreSchema = z.object({
  bucket: z.string().min(1),
  description: z.string().optional(),
  maxChunkSize: z.number().optional(),
  maxBytes: z.number().optional(),
  storage: z.enum(['file', 'memory']).optional(),
  replicas: z.number().optional(),
  ttl: z.number().optional(),
  connId: z.string().optional(),
});

objectStoreRouter.post('/objectstore', async (req, res) => {
  try {
    const config = objStoreSchema.parse(req.body);
    const connId = (req.query.connId || config.connId) as string;
    if (!connId) return res.status(400).json({ error: 'connId required' });
    const js = await getJetStream(connId);

    const opts: any = {};
    if (config.description) opts.description = config.description;
    if (config.maxChunkSize) opts.max_chunk_size = config.maxChunkSize;
    if (config.maxBytes) opts.max_bytes = config.maxBytes;
    if (config.storage) opts.storage = config.storage;
    if (config.replicas) opts.num_replicas = config.replicas;
    if (config.ttl) opts.ttl = config.ttl;

    const os = await js.views.os(config.bucket, opts);
    res.json({ success: true, bucket: config.bucket });
  } catch (err: any) {
    res.status(500).json({ error: err.message });
  }
});

objectStoreRouter.get('/objectstore/:store', async (req, res) => {
  try {
    const connId = req.query.connId as string;
    if (!connId) return res.status(400).json({ error: 'connId required' });
    const js = await getJetStream(connId);
    const os = await js.views.os(req.params.store);

    const objects: any[] = [];
    const list = await os.list();
    for await (const obj of list) {
      objects.push({
        name: obj.name,
        description: obj.description,
        size: obj.size,
        chunks: obj.chunks,
        nuid: obj.nuid,
        digest: obj.digest,
        deleted: obj.deleted,
        mtime: obj.mtime,
      });
    }
    res.json(objects);
  } catch (err: any) {
    res.status(500).json({ error: err.message });
  }
});

objectStoreRouter.get('/objectstore/:store/:name', async (req, res) => {
  try {
    const connId = req.query.connId as string;
    if (!connId) return res.status(400).json({ error: 'connId required' });
    const js = await getJetStream(connId);
    const os = await js.views.os(req.params.store);

    const result = await os.get(req.params.name);
    if (!result) {
      return res.status(404).json({ error: 'Object not found' });
    }

    // Read the data from the readable stream
    const chunks: Uint8Array[] = [];
    const reader = result.data.getReader();
    while (true) {
      const { done, value } = await reader.read();
      if (done) break;
      if (value) chunks.push(value);
    }

    const data = Buffer.concat(chunks);
    res.set('Content-Type', 'application/octet-stream');
    res.set('Content-Disposition', `attachment; filename="${req.params.name}"`);
    res.send(data);
  } catch (err: any) {
    res.status(500).json({ error: err.message });
  }
});

// Upload object - accepts JSON with base64-encoded data
objectStoreRouter.put('/objectstore/:store/:name', async (req, res) => {
  try {
    const connId = (req.query.connId || req.body?.connId) as string;
    if (!connId) return res.status(400).json({ error: 'connId required' });
    const nc = connectionStore.getConnection(connId);
    if (!nc) return res.status(400).json({ error: 'Not connected' });
    const js = nc.jetstream();
    const os = await js.views.os(req.params.store);

    const { data, description } = req.body;
    if (!data) return res.status(400).json({ error: 'data (base64) required' });

    const buffer = Buffer.from(data, 'base64');
    const stream = new ReadableStream({
      start(controller) {
        controller.enqueue(new Uint8Array(buffer));
        controller.close();
      }
    });

    const putOpts: any = { name: req.params.name };
    if (description) putOpts.description = description;

    const info = await os.put(putOpts, stream);
    res.json({ success: true, name: req.params.name, size: info.size, chunks: info.chunks });
  } catch (err: any) {
    res.status(500).json({ error: err.message });
  }
});

objectStoreRouter.delete('/objectstore/:store/:name', async (req, res) => {
  try {
    const connId = req.query.connId as string;
    if (!connId) return res.status(400).json({ error: 'connId required' });
    const js = await getJetStream(connId);
    const os = await js.views.os(req.params.store);
    await os.delete(req.params.name);
    res.json({ success: true });
  } catch (err: any) {
    res.status(500).json({ error: err.message });
  }
});
